paperqa/clients/openalex.py

```diff
@@ -0,0 +1,237 @@
+) -> DocDetails | None:
+    """Get paper details from OpenAlex given a DOI or paper title.
+
+    Args:
+        session: The active session of the request.
+        doi: The DOI of the paper.
+        title: The title of the paper.
+        fields: Specific fields to include in the request.
+        title_similarity_threshold: The threshold for title similarity.
+
+    Returns:
+        The details of the document if found, otherwise None.
+
+    Raises:
+        ValueError: If neither DOI nor title is provided.
+        DOINotFoundError: If the paper cannot be found.
+    """
+    mailto = await get_openalex_mailto()
+    params = {"mailto": mailto} if mailto else {}
+
+    if doi is title is None:
+        raise ValueError("Either a DOI or title must be provided.")
+
+    url = f"{OPENALEX_BASE_URL}/works"
+    if doi:
+        # this looks wrong but it's now
+        # will compile to a relative url similar to:
+        # https://api.openalex.org/works/https://doi.org/10.7717/peerj.4375
+        url += f"/https://doi.org/{quote(doi, safe='')}"
+    elif title:
+        params["filter"] = f"title.search:{title}"
+
+    if fields:
+        params["select"] = ",".join(fields)
+
+    async with session.get(
+        url,
+        params=params,
+        timeout=aiohttp.ClientTimeout(OPENALEX_API_REQUEST_TIMEOUT),
+    ) as response:
+        try:
+            response.raise_for_status()
+            response_data = await response.json()
+        except (aiohttp.ClientResponseError, json.JSONDecodeError) as exc:
+            raise DOINotFoundError("Could not find paper given DOI/title.") from exc
+
+        if response_data.get("status") == "failed":
+            raise DOINotFoundError(
+                "OpenAlex API returned a failed status for the query."
+            )
+
+        results_data = response_data
+        if params.get("filter") is not None:
+            results_data = results_data["results"]
+            if len(results_data) == 0:
+                raise DOINotFoundError(
+                    "OpenAlex API did not return any items for the query."
+                )
+            results_data = results_data[0]
+
+        if (
+            doi is None
+            and title
+            and strings_similarity(results_data.get("title", ""), title)
+            < title_similarity_threshold
+        ):
+            raise DOINotFoundError(
+                f"OpenAlex results did not match for title {title!r}."
+            )
+        if doi and results_data.get("doi") != doi:
+            raise DOINotFoundError("DOI not found in OpenAlex")
+
+        return await parse_openalex_to_doc_details(results_data)
+
+
+async def parse_openalex_to_doc_details(message: dict[str, Any]) -> DocDetails:
+    """Parse OpenAlex API response to DocDetails.
+
+    Args:
+        message: The OpenAlex API response message.
+
+    Returns:
+        Parsed document details.
+    """
+    authors = [
+        authorship.get("raw_author_name")
```

Comment:
This `.get` if it returns `None`, won't the `reformat_name` blow up?

Reply:
you're right -- I was thinking this could never be `None` so I didn't put much thought into it originally. but it's safer regardless.

Added some safeguards. lmk